Add a tutorial on how to use aggregations with OpenSearch

anelook opened the pull request:
New tutorial will help our users create and run aggregations with OpenSearch.
This is logical continuation of the previous tutorial - https://developer.aiven.io/docs/products/opensearch/howto/opensearch-and-nodejs.html

lornajane commented:
Does this tutorial need to be a sequel to the previous one, or could it be useful standalone to someone with a bit of OpenSearch experience but not much with the aggregations aspect?

anelook commented:
> Does this tutorial need to be a sequel to the previous one, or could it be useful standalone to someone with a bit of OpenSearch experience but not much with the aggregations aspect?

Since I didn't want to repeat the instructions for playground setup, and because we analyse same dataset I positioned it as a sequel. 

An alternative I thought of would be to split the first tutorial into preparation step and the actual search experiments. Then those parts can be referenced independently.

lornajane reviewed (commented):
I picked up a few minor things and I have some high level comments - I'd like to review your next set of changes and do a language review as well before I approve. I think this is in good shape and the sample data set continues to give us some great examples. The beginning is a bit of a rough experience. Perhaps we can number the setup steps in the other article, and then say in this one to follow the first four (or however many) steps.  Alternatively, be clearer that the repository is required as the basis of the exercise.

anelook commented on docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst:
```diff
@@ -0,0 +1,792 @@
+You can also clone the final demo project from `GitHub repository <https://github.com/aiven/demo-open-search-node-js>`_. You'll find the code for aggregations in the `aggregate.js <https://github.com/aiven/demo-open-search-node-js/blob/main/aggregate.js>`_.
+
+
+About aggregations
```
Done: moved to a separate concept article.

anelook commented on docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst:
```diff
@@ -0,0 +1,792 @@
+
+.. code-block:: javascript
+
+    const { client, indexName: index } = require("./config");
```
I've introduced different files in the step of preparation of a playground. This hopefully helps.

anelook commented on docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst:
```diff
@@ -0,0 +1,792 @@
+
+You'll need an OpenSearch cluster and a NodeJS project. Make sure that you have `the necessary essentials <https://developer.aiven.io/docs/products/opensearch/howto/opensearch-and-nodejs.html#prerequisites>`_ in place, that you're `connected to the OpenSearch cluster <https://developer.aiven.io/docs/products/opensearch/howto/opensearch-and-nodejs.html#connect-to-opensearch>`_ and have `example da ...
+
+You can also clone the final demo project from `GitHub repository <https://github.com/aiven/demo-open-search-node-js>`_. You'll find the code for aggregations in the `aggregate.js <https://github.com/aiven/demo-open-search-node-js/blob/main/aggregate.js>`_.
```
Added proper details, so no step is missing. Playground setup is now a separate tutorial

anelook commented on docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst:
```diff
@@ -0,0 +1,792 @@
+     - and so on
+
+
+For every data point (a year in our case) we take the count of added recipes, add number of recipes added over last two years and divide the result by three (according to the size of our window). For the first and second year we divide by the number of available years (1 and 2 respectively). And this is how moving average is calculated. If you compare numbers from the table with the numbers retur ...
```
Happy that you could easily follow it! Pipeline aggregations are very tricky to explain. And recipe data isn't perfect for them. I want to write a separate article and use some log data instead.

anelook commented on docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst:
```diff
@@ -0,0 +1,792 @@
+
+::
+
+    run-func aggregate terms categories.keyword 3
```
Hehe, not sure how it sneaked in! Fixed

anelook commented on docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst:
```diff
@@ -0,0 +1,792 @@
+
+::
+
+    run-func index rangeSearch calories 3256
```
It can be unbounded, I added extra phrase about it

anelook commented:
Based on the feedback I revisited unclear chunks and made these improvements:

- Moved playground setup to a separate article, which is referenced by both Search and Aggregations tutorials.
- Moved description of aggregations into concept article
- Replaced Present Perfect tense with Past tense
- Clarified what code goes to what files. Instead of keeping all in a single index.js, introduced config.js, index.js, helpers.js, search.js and aggregation.js and explained their roles.
- Minor simplifications and alignment when it comes to code and commands.

anelook commented:
Direct links to added and modified content

Setup Playground: https://deploy-preview-310--developer-aiven-io-preview.netlify.app/docs/products/opensearch/howto/setup-opensearch-and-nodejs-playground.html

Search: https://deploy-preview-310--developer-aiven-io-preview.netlify.app/docs/products/opensearch/howto/opensearch-and-nodejs.html

Aggregations: https://deploy-preview-310--developer-aiven-io-preview.netlify.app/docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.html

Aggregation concepts: http://127.0.0.1:8000/docs/products/opensearch/concepts/aggregations.html

lornajane commented:
Hmm, I sort of like this but I think there are still some really useful pieces of documentation hiding in this very specific set of tutorials ... trying to think about the right structure for this is making my head hurt! There's a lot going on here, and I think we need to move some of these things out into reference articles in their own right that can be used by anyone to do these things for their own projects rather than in a grouped playground setup task.
 - creating an opensearch service. This should be in getting started, if it isn't we should fix that, and just link to it.
 - importing the data. We already have a sample data setup page. Can we direct people to this existing resource? Even if we add the Node equivalent instructions after the Python ones.
 - initialising a node project. We aren't teaching NodeJS so let's remove this section (but the overview of the files is useful to orientate for the search and aggregate articles)
 - basic use of `avn service` commands - I think this is out of scope when teaching someone about aggregations but it would be fine to add a tip linking to the CLI docs
 - Connecting to NodeJS - this should be a code sample, people will want this outside of these specific articles.

I think the PostgreSQL section is a good illustration of how I was thinking about this content. There's a post about getting started, one about sample data, some connection examples, and then some other pieces that refer back to these "building blocks" - but each of the blocks are standalone and discoverable.

anelook commented:
Added changes after our conversation about the structure.

lornajane commented:
@anelook This is in good shape! I've added some language edits that you might want to double-check, and there's a small conflict.

anelook commented:
Thank you for the language improvements! All looks good to me. 

I've pushed a rebased version and GitHub demands a new review.

anelook commented:
I've rebased the branch, it is up to date in relation to ``main``.

lornajane commented on _toc.yml:
```diff
@@ -59,6 +60,7 @@ entries:
         title: HowTo
         entries:
           - file: docs/products/kafka/howto/kafka-sasl-auth
+          - file: docs/products/kafka/howto/integrate-service-logs-into-kafka-topic
```
I'm not sure what this change is doing here!

anelook replied:
😨 This was not supposed to be there! Thx for catching it. Removed

lornajane reviewed (approved):
Great job, I know it was a long road!
